module/zfs/dmu.c

```diff
@@ -1040,100 +1269,726 @@ dmu_read_by_dnode(dnode_t *dn, uint64_t offset, uint64_t size, void *buf,
+		db->db_buf = NULL;
+		db->db.db_data = NULL;
+		dr->dt.dl.dr_data = NULL;
+		/*
+		 * Destroy the data buffer if it is not in use.
+		 */
+		for (dr = list_head(&db->db_dirty_records);
+		    dr != NULL;
+		    dr = list_next(&db->db_dirty_records, dr)) {
+			if (dr->dt.dl.dr_data == buf)
+				break;
+		}
+		if (dr == NULL)
+			arc_buf_destroy(buf, db);
+	}
+	ASSERT(db->db.db_data == NULL);
+	db->db_state = DB_UNCACHED;
+	mutex_exit(&db->db_mtx);
+
+	dmu_sync_done(zio, NULL, zio->io_private);
+	kmem_free(zio->io_bp, sizeof (blkptr_t));
+}
+
+static int
+dmu_write_direct(zio_t *pio, dmu_buf_impl_t *db, abd_t *data, dmu_tx_t *tx)
+{
+	objset_t *os = db->db_objset;
+	dsl_dataset_t *ds = os->os_dsl_dataset;
+	dbuf_dirty_record_t *dr_head, *dr_next;
+	dmu_sync_arg_t *dsa;
+	zbookmark_phys_t zb;
+	zio_prop_t zp;
+	dnode_t *dn;
+	uint64_t txg = dmu_tx_get_txg(tx);
+	blkptr_t *bp;
+	zio_t *zio;
+	int err = 0;
+
+	ASSERT(tx != NULL);
+
+	SET_BOOKMARK(&zb, ds->ds_object,
+	    db->db.db_object, db->db_level, db->db_blkid);
+
+	/*
+	 * No support for this
+	 */
+	if (txg > spa_freeze_txg(os->os_spa))
+		return (SET_ERROR(ENOTSUP));
+
+	DB_DNODE_ENTER(db);
+	dn = DB_DNODE(db);
+	dmu_write_policy(os, dn, db->db_level, WP_DMU_SYNC, &zp);
+	DB_DNODE_EXIT(db);
+
+	/*
+	 * Dirty this dbuf with DB_NOFILL since we will not have any data
+	 * associated with the dbuf.
+	 */
+	dmu_buf_will_not_fill(&db->db, tx);
+
+	/* XXX - probably don't need this, since we are in an open tx */
+	mutex_enter(&db->db_mtx);
+
+	ASSERT(txg > spa_last_synced_txg(os->os_spa));
+	ASSERT(txg > spa_syncing_txg(os->os_spa));
+
+	dr_head = list_head(&db->db_dirty_records);
+	dr_next = list_next(&db->db_dirty_records, dr_head);
+	VERIFY(dr_head->dr_txg == txg);
+
+	bp = kmem_alloc(sizeof (blkptr_t), KM_SLEEP);
+	if (db->db_blkptr != NULL) {
+		/*
+		 * fill in bp with current blkptr so that
+		 * the nopwrite code can check if we're writing the same
+		 * data that's already on disk.
+		 */
+		*bp = *db->db_blkptr;
+	} else {
+		bzero(bp, sizeof (blkptr_t));
+	}
+
+	/*
+	 * Disable nopwrite if the current BP could change before
+	 * this TXG syncs.
+	 */
+	if (dr_next != NULL)
+		zp.zp_nopwrite = B_FALSE;
+
+	ASSERT(dr_head->dt.dl.dr_override_state == DR_NOT_OVERRIDDEN);
+	dr_head->dt.dl.dr_override_state = DR_IN_DMU_SYNC;
+	mutex_exit(&db->db_mtx);
+
+	/*
+	 * We will not be writing this block in syncing context, so
+	 * update the dirty space accounting.
+	 * XXX - this should be handled as part of will_not_fill()
+	 */
+	dsl_pool_undirty_space(dmu_objset_pool(os), dr_head->dr_accounted, txg);
+
+	dsa = kmem_alloc(sizeof (dmu_sync_arg_t), KM_SLEEP);
+	dsa->dsa_dr = dr_head;
+	dsa->dsa_done = NULL;
+	dsa->dsa_zgd = NULL;
+	dsa->dsa_tx = NULL;
+
+	zio = zio_write(pio, os->os_spa, txg, bp, data,
+	    db->db.db_size, db->db.db_size, &zp,
+	    dmu_write_direct_ready, NULL, NULL, dmu_write_direct_done, dsa,
+	    ZIO_PRIORITY_SYNC_WRITE, ZIO_FLAG_CANFAIL, &zb);
```

Comment:
I'm a little concerned about bypassing `zio_dva_throttle()`. Background: [slides 11-16](https://www.bsdcan.org/2016/schedule/attachments/366_ZFS%20Allocation%20Performance.pdf) and [video](https://youtu.be/bzqlzjbzyGA?t=1297) from my talk at BSDCAN 2016.


 This means that DIRECTIO writes will be spread out among the vdevs using the old round-robin algorithm.  This could potentially result in poor performance due to allocating from the slowest / most fragmented vdev, and we could potentially make the vdevs even more imbalanced (at least in terms of performance/fragmentation).  @grwilson do you have any thoughts on this? How big the impact could be, and potential ways to mitigate?  Could we make this use the throttle?